### DISCUSSION TITLE
Stress offset for water

### DISCUSSION THREAD

**Discussion by emilfrost** [ACEsuit/mace#949] [Q&A] [answered]
Hi

I have observed an offset in the diagonal part of the stress tensor when training on bulk water that seems to be present both when training from scratch and when finetuning a foundations model. It is possible to eliminate the offset by increasing the weight on stress in the loss function during training, but there seems to be a trade-off (inverse relation) between energy and stress performance. 

I have attached a training run that illustrates the problem. This is from (naive) finetuning of the MPA-0 model. The stage one weights are 10:1:10 for forces, energy, and stress, respectively. The stage two weights are 1:10:10, in the same order. The RMSEs for the energy and stress are strongly correlated. Looking at the stress parity plots, most of the points are offset from the y=x line (these are the diagonal entries) while some stay on the line (these are the off-diagonal entries). For completeness, here is the config file: [config.txt](https://github.com/user-attachments/files/20148109/config.txt)

![ft_naive_mpa_0_run-123_train_rpbe_stage_one](https://github.com/user-attachments/assets/04537e55-7bda-4c93-b693-0081d93940c5)
![ft_naive_mpa_0_run-123_train_rpbe_stage_two](https://github.com/user-attachments/assets/ecdb4e30-a6fc-46ec-85a9-731d9a129591)

**For context**
I'm training a MACE model on a very simple dataset consisting of around 200 periodic boxes of water with 27 H2O molecules sampled from -10% to 20% volume change from equilibrium and temperatures between 300K and 500K. The data has been sampled from MD simulations (NVT) with a foundation model and subsequently evaluated with the RPBE functional in VASP. I would like to use the model for NPT ensembles but a diagonal stress offset gives a systematic density error.

Is this a well-known problem and do you have any suggestions to alleviate the problem other than turning up the stress weight? Is it possible to make the stress prediction more flexible to not compromise energies?

Emil

**Comment by gabor1** [comment, marked as answer]
Unless I've got my unit conversions wrong, the offset you are talking about is about 0.1 GPa. It's not zero but very small. Are you sure that your training data is converged to this level? Stresses converge quite slowly with kpoints and cutoff.

**Reply by emilfrost** [reply]
I have not done a convergence check on the stress yet but I have run some NPT ensembles to check the water density. I used a 3x3x3 water cell at 300K, 1 atm so it should be close to the training set. The offset seems to change the density by about 0.1 g/cm^3.

Fitting a straight line to the stress parity plot, the offset is 1.45 meV/Å^3 for the diagonal entries. I have attached below two NPT simulations with the stage two model: one with default stress and one where I've manually added the shift to the diagonal. First part is a short NVT run to equilibrate the temperature and then 100ps of NPT where the pressure (blue) approaches and fluctuates around 1 atm \approx 0 eV/Å^3. The offset gives rise to a difference in water density (orange) of about 0.1 g/cm^3. Adding D3 corrections on top then gives approximately 1 g/cm^3 in density.

![pressure_density](https://github.com/user-attachments/assets/4e3bbfe1-b8f7-4a91-8f82-1c2cd465258e)
![pressure_density_corrected](https://github.com/user-attachments/assets/a73c9e59-19de-443b-add9-dfbe5d3feef8)

**Reply by emilfrost** [reply]
As a follow-up, here's a quick back-of-the-envelope calculation. The compressibility of water is

$\beta =-\frac{1}{V} \frac{\partial{V}}{\partial{p}}=4.6 \times 10^{-10} \mathrm{Pa^{-1}}$,

meaning the change in volume given a (small) change in pressure is (assuming constant compressibility)

$\Delta V \approx -\beta V \Delta p$.

The resulting change in relative density is then $\frac{\Delta \rho}{\rho} \approx - \frac{\Delta V}{V} = \beta \Delta p = 4.6 \times 10^{-10} \mathrm{Pa^{-1}} \cdot 1.45 \mathrm{meV/Å^3} = 0.11 $.

For water, such an offset roughly corresponds to 10% density change.

**Reply by gabor1** [reply]
I agree - which is why I was questioning the accuracy of your point wise stresses in your data. that's not the same as converging the density with your ab initio MD. plot the stress of one of your configs as a function of cutoff and kpoint.

**Reply by gabor1** [reply]
Actually 1.6e10 (Pa per eV/A^3) * 0.00145 (eV/A^3) * 4.6*1e-10 (1/Pa, compressibility) =0.0107 , so you appear to be off by a factor of 10.

**Reply by emilfrost** [reply]
My point was just that even such a small offset would cause noticeable changes in the density. Also, I think the conversion factor you mentioned is 1.6e11 Pa = 1 eV/A^3, so the 11% should be correct.

Anyway, here's the convergence test for one of the configurations from the test set. It shows the difference to the calculation with the highest energy cutoff and number of k-points. 

The convergence with respect to k-points is very fast, but not with respect to energy cutoff. I shaded the plots according to good MACE errors one might see. I used 1 meV/atom for energy uncertainty, 10 meV/A for forces, and 0.2 meV/Å^3=0.03 GPa for stress components. I think these errors are realistic in systems of bulk water. I didn't know the k-point convergence would be so fast, so I went a bit overboard - all curves basically lie on top of each other. A large energy cutoff still runs fast, when so few k-points are needed so that's nice. You could probably get away with just using the Gamma point for boxes of water, but at least a density of 20 Å should be enough. 

I'm surprised to see the diagonal components of the stress tensor need such a large energy cutoff (1000 eV). Before, I used 500 eV (and 50Å k-point density) which gives an error very similar to the offset I'm seeing (0.3 GPa = 1.9 meV/Å^3). Let me redo the calculations with these converged settings (ENCUT=1000eV, k-point-density=20Å) and see if MACE is more willing to fit the data.

![energy_convergence](https://github.com/user-attachments/assets/55d84312-58c4-48ab-ac18-5ae79ee6ee15)
![force_convergence](https://github.com/user-attachments/assets/391463c9-9be1-4e81-98f5-a8f4a338f7bd)
![stress_convergence](https://github.com/user-attachments/assets/8753ca03-e7e0-4a94-bf4a-2b70a2151c17)

**Reply by gabor1** [reply]
1 pa  =  6.241509e-11 ev/A3; 1 ev/A3  =  1.602177e+10 pa

**Reply by gabor1** [reply]
Also, if you are looking for a MACE accuracy of eps, I think it would be good to have data that is reliably converged to better than eps, perhaps eps/2 or eps/5...

**Reply by emilfrost** [reply]
It's working really well now and the stress offset is now negligible (factor 10 smaller). This is from a run with the default weights during training. There are some configurations in the test set with the old VASP settings - please disregard those outliers in the parity plots.

Thanks a lot for the help, it was definitely a data problem.

![ft_naive_matpes_run-123_train_rpbe_stage_two](https://github.com/user-attachments/assets/74b5a66c-cf67-416a-a36e-773111953596)
